rfd/0147-repo-fork-syncing.md

```diff
@@ -0,0 +1,60 @@
+ 
+## Details
+ 
+As such we need a solution that can provide the following:
+* Automatic and regular updates to a repo to ensure it's in sync
+* The ability to apply changes to the upstream repo (for example disable Actions that we don't want running on the repo copy, or enable additional scanning / actions)
+ 
+### Branch Structure
+ 
+* `main` - This will be the repos default branch.  This branch will contain the upstream history in addition to commits needed for the sync process or other custom changes on the repo copy
+* `sync/upstream-[UPSTREAM_BRANCH_NAME]` - This branch will be an exact copy of the upstream branch.  If there is a need to create a PR in this repo the history should be based off this rather than `main`
+* `sync/rebase` - This branch is where our custom changes will be committed.  It will be rebased on to the upstream changes before being committed into `main`
+ 
+#### Branch Protections
+ 
+The only branch that will be protected is the `sync/rebase` branch.  Other branches are expected to allow force pushes so they can be forcefully synced based off the upstream state.
+ 
+### High Level Implementation
+ 
+At a high level there are two primary components to this implementation:
+1. A GitHub action which will be scheduled as well as triggered on changes to the `sync/rebase` branch.  This action will perform the actual rebase action and commit the results to the respective branches above.
+2. A script that is invoked when the rebase can not be applied cleanly.  This can allow some custom rebase logic as a means to reduce how often the `sync/rebase` branch needs to be manually updated to avoid rebase conflicts.  A simple example is that workflow automations not desired in the repo copy may be removed in the `sync/rebase` branch.  That means that any changes to those committed workflows would result in a conflict.  However it can be safely resolved that any modified files removed in the `sync/rebase` branch can be resolved by simply removing.
+ 
+At a high level the action will do the following steps:
+1. Setup auth and any necessary repo specific configuration
+2. Checkout the latest upstream contents
+3. Fetch our custom changes
+4. Push the exact copy branches `sync/upstream-[UPSTREAM_BRANCH_NAME]` to our fork
+5. Rebase our `sync/rebase` branch on to the latest upstream changes, using the script for resolutions if necessary and possible
+6. Force push the rebased changes to the `main` branch on our repo copy
+ 
+### Teleport Security Scanning
+ 
+This tooling will be leveraged to address the gaps in our release branch scanning by doing the following:
+1. Create 3 repos (1 for each supported version): `teleport-sec_scan-1`, `teleport-sec_scan-2`, `teleport-sec_scan-3`.  `1` will reference our latest release branch, with each subsequent version inspecting a consecutively older version.
```

Comment:
So we'll be creating 3 or 4 _new repos_ per year that need to be kept in sync and have actions running against them? That seems costly both from a time and management perspective as well as for all the automation that will need to run.

Also, isn't there some redundancy here? It seems for each supported version, we'll have both:

- a sync/upstream branch in a private fork that's kept in sync
- a dedicated sec_scan repo that's also kept in sync

That's a lot of duplication, don't you think? Any way we can combine them?

Also, pretty please don't mix `_` and `-` in the repo names. It's begging for typos and also just looks awkward IMO 🙂

Reply:
> per year 

This is incorrect.  The branch which the repo is syncing will automatically be updated based off the patterns within the upstream repo.  So `teleport-sec_scan-1` will always point to the most recent release version branch.

> isn't there some redundancy

There is, but it's due to different needs.  With `teleport-private` we are focused on improving the developer experiencing for crafting security fixes.  With these `teleport-sec_scan` repos we are working around a deficiency in Dependabot where only one branch per repo can be scanned (the default branch).